src/shared/dissect-image.c

```diff
@@ -2682,6 +2689,29 @@ int verity_dissect_and_mount(const char *src, const char *dest, const MountOptio
         if (r < 0)
                 return log_debug_errno(r, "Failed to mount image: %m");
 
+        /* If we got os-release values from the caller, then we need to match them with the image's
+         * extension-release.d/ content. Return -EINVAL if there's any mismatch.
+         * First, check the distro ID. If that matches, then check the new SYSEXT_LEVEL value if
+         * available, or else fallback to VERSION_ID. */
+        if (host_os_release_id && (host_os_release_version_id || host_os_release_sysext_level)) {
+                _cleanup_strv_free_ char **extension_release = NULL;
+
+                r = load_extension_release_pairs(dest, dissected_image->image_name, &extension_release);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse image %s extension-release metadata: %m", dissected_image->image_name);
+
+                r = extension_release_validate(
+                        dissected_image->image_name,
+                        host_os_release_id,
+                        host_os_release_version_id,
+                        host_os_release_sysext_level,
+                        extension_release);
+                if (r == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Image %s extension-release metadata does not match the root's", dissected_image->image_name);
```

Comment:
maybe use some recognizable error message for this. ESTALE? EBADSLT? EDOM?

Reply:
changed to ESTALE